# Estructuras Dataframe

In [1]:
import pandas as pd

In [2]:
vuelos=pd.read_csv('DelayedFlights.csv')
vuelos

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0         2211.0        2225            WN  ...     4.0     8.0          0   
1         1002.0        1000            WN  ...     5.0    10.0          0   
2          804.0         750            WN  ...     3.0    17.0          0   
3         1959.0        1925            WN  ...     3.0    10.0          0   
4         2121.0        2110            WN  ...     4.0    10.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1936753   1617.0        1552            DL  ...     9.0    18.0          0   
1936754    904.0         749            DL  ...    15.0    34.0          0   
1936755   1149.0        1010            DL  ...     8.0    32.0          0   
1936756   1446.0        1437            DL  ...    13.0    13.0          0   
1936757   1413.0        1418            DL  ...     8.0    11.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                       N         0           NaN           NaN      NaN   
1                       N         0           NaN           NaN      NaN   
2                       N         0           NaN           NaN      NaN   
3                       N         0           2.0           0.0      0.0   
4                       N         0           NaN           NaN      NaN   
...                   ...       ...           ...           ...      ...   
1936753                 N         0           3.0           0.0      0.0   
1936754                 N         0           0.0          57.0     18.0   
1936755                 N         0           1.0           0.0     19.0   
1936756                 N         0           NaN           NaN      NaN   
1936757                 N         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
0                 NaN                NaN  
1                 NaN                NaN  
2                 NaN                NaN  
3                 0.0               32.0  
4                 NaN                NaN  
...               ...                ...  
1936753           0.0               22.0  
1936754           0.0                0.0  
1936755           0.0               79.0  
1936756           NaN                NaN  
1936757           NaN                NaN  

[1936758 rows x 30 columns]

In [3]:
#Reducción a 200000 registros (filas) para trabajar en todos los ejercicios

vuelos=vuelos.sample(200000)

# Ejercicio 1. 
## Visualización de todas las variables existentes en el DataFrame para la elección de las que tengan información relevante para su análisis estadístico, justificándolo.

In [4]:
# Comprbación de las variables (o columnas) existentes.

vuelos.columns

Index(['Unnamed: 0', 'Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime',
       'CRSDepTime', 'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum',
       'TailNum', 'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn', 'TaxiOut',
       'Cancelled', 'CancellationCode', 'Diverted', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'],
      dtype='object')

### Información sobre el significado de cada variable

"Year":
año del vuelo (todos del 2008).

"Month":
mes del vuelo.

"DayOfMonth":
día del mes (1 a 31).

"DayOfWeek":
día de la semana.

"DepTime":
hora de salida real.

"CRSDepTime":
hora de salida programada.

"ArrTime":
hora de llegada real.

"CRSArrTime":
hora de llegada programada.

"UniqueCarrier":
ID de operador.

"FlightNum":
número de vuelo.

"TailNum":
número de cola del avión.

"ActualElapsedTime":
tiempo real transcurrido del vuelo, en minutos.

"CRSElapsedTime":
tiempo transcurrido programado del vuelo, en minutos.

"AirTime":
tiempo en el aire para el vuelo, en minutos.

"ArrDelay":
retardo de llegada, en minutos.

"DepDelay":
retraso de salida, en minutos.

"Origin":
aeropuerto de origen.

"Dest":
aeropuerto de destino.

"Distance":
distancia de vuelo, en millas.

"TaxiIn":
tiempo de rodaje desde las ruedas hasta la llegada a la puerta, en minutos.

"TaxiOut":
tiempo de rodaje desde la salida de la puerta hasta las ruedas, en minutos.

"Cancelled":
estado de cancelación, donde Sí=1 y No=0.

"CancellationCode":
código de cancelación, en caso de existir.

"Diverted":
estado de desvío, donde Sí=1 y No=0.

"CarrierDelay":
retardo, en minutos, atribuible al transportista.

"WeatherDelay":
retraso, en minutos, atribuible a factores meteorológicos.

"NASDelay":
retraso, en minutos, atribuible al Sistema Nacional de Aviación.

"SecurityDelay":
retardo, en minutos, atribuible a factores de seguridad.

"LateAircraftDelay":
retraso, en minutos, atribuible a aeronaves que llegan tarde.

### Eliminación de columnas

Como el objetivo del nuevo DataFrame es su análisis estadístico sobre los tiempos de los vuelos, he eliminado las variables (o columnas) que carecen de tal fin, tales como:

'Unnamed: 0', aparentemente tiene la forma de un indice para el dataset, aunque no tiene un orden consecutivo. Aún así, pandas ya generará un indice ordenado de números para cada registro; 

'Year', ya que todos los registros corresponden al año 2008; 

'FlightNUm' y 'TailNum', ambas solo contienen información de interés al ámbito de control interno para las compañías aéreas;

'TaxiIn' y 'TaxiOut', ya están dentro del tiempo de vuelo;

'CancellationCode', lo que interesa es disponer de si existe cancelación o no del vuelo (algo con lo que el dataframe ya cuenta);

'CarrierDelay', 'WeatherDelay', 'NASDelay', ' 'SecurityDelay' y 'LateAircraftDelay', aquí no analizamos la responsabilidad por el retraso del vuelo, sino los tiempos que se han empleado (además, muchos de sus valores carecen de registro).

### En consecuencia, el nuevo dataframe quedaría de la siguiente manera:

In [5]:
vuelos.drop(['Unnamed: 0', 'Year', 'FlightNum', 'TailNum', 'TaxiIn', 'TaxiOut', 'CancellationCode', 
             'CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'], axis=1, inplace=True)

## Ejercicio 2

In [6]:
# a) Resumen estadístico y análisis de datos relevantes (no habituales o evidentes)
vuelos[['CRSDepTime', 'ActualElapsedTime',
        'AirTime', 'DepDelay', 'ArrDelay', 'Distance']].describe().round(2)

CRSDepTime  ActualElapsedTime    AirTime   DepDelay   ArrDelay  \
count   200000.00          199181.00  199181.00  200000.00  199181.00   
mean      1467.90             133.37     108.37      43.18      42.22   
std        425.32              72.07      68.67      53.31      56.71   
min          5.00              16.00       1.00       6.00     -66.00   
25%       1135.00              80.00      58.00      12.00       9.00   
50%       1510.00             116.00      90.00      24.00      25.00   
75%       1817.00             165.00     137.00      54.00      56.00   
max       2359.00             727.00     651.00    1292.00    1525.00   

        Distance  
count  200000.00  
mean      766.55  
std       574.75  
min        30.00  
25%       340.00  
50%       607.00  
75%       998.00  
max      4962.00

Como puede observarse, la mitad de los vuelos se efectúan pasado poco más el mediodía (unos minutos después de las 15:00), y solo tres horas más tarde (alrededor de las 18:00) ya se han realizado el 75%.

Por otra parte, la media del transcurso real de los vuelos es de 133.37 minutos.

También llama la atención el vuelo que emplea solo un minuto en el aire (* en la siguiente línea podrá verse tal/es registro/s), Como es lógico, se trata de algún fallo en el mismo, mas no de un 'outlier', ya que el la distancia de estos vuelos superan los 461 millas, incluso por encimma de las 700.

El mismo tiempo de retraso de las salidas se traslada al de las llegadas, así se puede ver en sus respectivas medias, las cuales son prácticamente idénticas.

Los datos registrados en los retrasos de llegada son verdaderamente extremos, hay mucha dispersión. Así lo reflejan sus desviaciones típicas, las cuales están incluso por encima de sus respectivas medias. Destacando en los retrasos de llegada un valor en negativo, lo que supone un adelanto en la llegada con respecto a la hora programada de 66 minutos (**).

Por último, cabe resaltar la distancia máxima de 4962 kms y la mínima de 30 kms, también con una desviación típida muy extensa, y con una media de 766.04 kms recorridos por cada vuelo.

In [10]:
# * visualización del resgistro con valor de 1 minuto en el tiempo de vuelo
vuelos[vuelos.AirTime==1]


Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
1248647      7          29          2   2235.0        2005    139.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  \
1248647        2146            B6              184.0           101.0      1.0   

         ArrDelay  DepDelay Origin Dest  Distance  Cancelled  Diverted  
1248647     233.0     150.0    JFK  PWM       274          0         0

In [12]:
# ** visualización del registro con máximo adelanto de la llegada
vuelos[vuelos.ArrDelay==-66]

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
50752      1          27          7   1743.0        1735   2049.0        2155   
50338      1          23          3   1708.0        1701   1805.0        1911   

      UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  \
50752            OH              246.0           320.0    205.0     -66.0   
50338            OH               57.0           130.0     25.0     -66.0   

       DepDelay Origin Dest  Distance  Cancelled  Diverted  
50752       8.0    JFK  AUS      1522          0         0  
50338       7.0    LGA  CMH       478          0         0

### b) Datos faltantes por columna


In [13]:
vuelos.isna().sum()


Month                  0
DayofMonth             0
DayOfWeek              0
DepTime                0
CRSDepTime             0
ArrTime              690
CRSArrTime             0
UniqueCarrier          0
ActualElapsedTime    819
CRSElapsedTime        12
AirTime              819
ArrDelay             819
DepDelay               0
Origin                 0
Dest                   0
Distance               0
Cancelled              0
Diverted               0
dtype: int64

In [14]:
print('Si sumamos todos los datos faltantes, hay un total de', vuelos.isna().sum().sum(), 'valores carentes de contenido.')

Si sumamos todos los datos faltantes, hay un total de 3159 valores carentes de contenido.


#### Eliminación de las filas con datos nulos

In [15]:
#Comprobación de los registros (o filas) con valores nulos

filas_nulos=vuelos[vuelos.isna().any(axis=1)]
filas_nulos

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
1235694      7          30          3   1719.0        1650      NaN   
1544058     10          29          3   1632.0        1620   2007.0   
968792       6           2          1   2221.0        2205      NaN   
1715803     11          10          1   1752.0        1735   2324.0   
22712        1          24          4   1449.0        1430      NaN   
...        ...         ...        ...      ...         ...      ...   
676449       4           8          2   2238.0        1835      NaN   
563939       3           4          2   1812.0        1802      NaN   
987260       6          27          5   1448.0        1440      NaN   
1760611     12          23          2    632.0         615   1436.0   
1809280     12          13          6   1507.0        1120   2012.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  \
1235694        1825            AA                NaN           155.0      NaN   
1544058        1845            WN                NaN           265.0      NaN   
968792         2346            UA                NaN           221.0      NaN   
1715803        2055            AA                NaN           260.0      NaN   
22712          1700            WN                NaN           150.0      NaN   
...             ...           ...                ...             ...      ...   
676449         2029            MQ                NaN            54.0      NaN   
563939         2010            CO                NaN           128.0      NaN   
987260         1709            DL                NaN           149.0      NaN   
1760611        1210            WN                NaN           235.0      NaN   
1809280        1310            OO                NaN           110.0      NaN   

         ArrDelay  DepDelay Origin Dest  Distance  Cancelled  Diverted  
1235694       NaN      29.0    JAX  DFW       919          0         1  
1544058       NaN      12.0    MDW  SAN      1728          0         1  
968792        NaN      16.0    IAD  DEN      1452          0         1  
1715803       NaN      17.0    JFK  DFW      1391          0         1  
22712         NaN      19.0    TPA  ISP      1034          0         1  
...           ...       ...    ...  ...       ...        ...       ...  
676449        NaN     243.0    GRB  ORD       174          0         1  
563939        NaN      10.0    BOS  CLE       563          0         1  
987260        NaN       8.0    ATL  LGA       761          0         1  
1760611       NaN      17.0    SEA  MDW      1733          0         1  
1809280       NaN     227.0    DEN  BZN       525          0         1  

[819 rows x 18 columns]

In [16]:
vuelos.isna().any(axis=1).sum()

819

Como la cantidad de las filas con datos nulos son prácticamente insignificantes en proporción a la cantidad de filas totales (no llega ni al 0.5%), y con la finalidad de mantener todos los registros, la cancelación de estos valores no inlfuirá en análisis de cualquier tipo que se realice sobre el dataset.

In [17]:
# Eliminación de datos nulos y comprobación del dataframe ahora sin las filas que los contenían

vuelos.dropna()

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
893067       6          10          2   1408.0        1255   1515.0   
869379       5          27          2   1026.0        1020   1614.0   
303381       2          11          1   1052.0        1045   1201.0   
1619984     10          23          4   1803.0        1750   2011.0   
474335       3          21          5   2209.0        2125   2251.0   
...        ...         ...        ...      ...         ...      ...   
888638       6           7          6   1202.0        1155   1413.0   
302880       2           5          2    932.0         810   1650.0   
1306003      8          15          5   1657.0        1648   2042.0   
928609       6          30          1   1331.0        1305   1546.0   
851564       5           9          5   1116.0        1055   1330.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  \
893067         1430            WN              127.0           155.0    112.0   
869379         1628            B6              228.0           248.0    195.0   
303381         1200            MQ               69.0            75.0     48.0   
1619984        2009            CO              248.0           259.0    221.0   
474335         2217            DL              102.0           112.0     81.0   
...             ...           ...                ...             ...      ...   
888638         1405            WN              131.0           130.0    115.0   
302880         1525            HA              318.0           315.0    282.0   
1306003        2018            YV              165.0           150.0    117.0   
928609         1512            XE              195.0           187.0    173.0   
851564         1255            AA              194.0           180.0    150.0   

         ArrDelay  DepDelay Origin Dest  Distance  Cancelled  Diverted  
893067       45.0      73.0    STL  DEN       770          0         0  
869379      -14.0       6.0    LGB  ORD      1735          0         0  
303381        1.0       7.0    LAX  LAS       236          0         0  
1619984       2.0      13.0    IAH  SFO      1635          0         0  
474335       34.0      44.0    SLC  SMF       532          0         0  
...           ...       ...    ...  ...       ...        ...       ...  
888638        8.0       7.0    TPA  IAD       810          0         0  
302880       85.0      82.0    HNL  LAX      2556          0         0  
1306003      24.0       9.0    COS  ORD       911          0         0  
928609       34.0      26.0    EWR  MSY      1167          0         0  
851564       35.0      21.0    LGA  MSP      1020          0         0  

[199181 rows x 18 columns]

### c) Creación de nuevas columnas

#### Velocidad media por vuelo

In [18]:
vuelos['AvarageSpeed']= (vuelos['Distance'] / vuelos['AirTime']*60).round(2)
vuelos

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
893067       6          10          2   1408.0        1255   1515.0   
869379       5          27          2   1026.0        1020   1614.0   
303381       2          11          1   1052.0        1045   1201.0   
1619984     10          23          4   1803.0        1750   2011.0   
474335       3          21          5   2209.0        2125   2251.0   
...        ...         ...        ...      ...         ...      ...   
888638       6           7          6   1202.0        1155   1413.0   
302880       2           5          2    932.0         810   1650.0   
1306003      8          15          5   1657.0        1648   2042.0   
928609       6          30          1   1331.0        1305   1546.0   
851564       5           9          5   1116.0        1055   1330.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  \
893067         1430            WN              127.0           155.0    112.0   
869379         1628            B6              228.0           248.0    195.0   
303381         1200            MQ               69.0            75.0     48.0   
1619984        2009            CO              248.0           259.0    221.0   
474335         2217            DL              102.0           112.0     81.0   
...             ...           ...                ...             ...      ...   
888638         1405            WN              131.0           130.0    115.0   
302880         1525            HA              318.0           315.0    282.0   
1306003        2018            YV              165.0           150.0    117.0   
928609         1512            XE              195.0           187.0    173.0   
851564         1255            AA              194.0           180.0    150.0   

         ArrDelay  DepDelay Origin Dest  Distance  Cancelled  Diverted  \
893067       45.0      73.0    STL  DEN       770          0         0   
869379      -14.0       6.0    LGB  ORD      1735          0         0   
303381        1.0       7.0    LAX  LAS       236          0         0   
1619984       2.0      13.0    IAH  SFO      1635          0         0   
474335       34.0      44.0    SLC  SMF       532          0         0   
...           ...       ...    ...  ...       ...        ...       ...   
888638        8.0       7.0    TPA  IAD       810          0         0   
302880       85.0      82.0    HNL  LAX      2556          0         0   
1306003      24.0       9.0    COS  ORD       911          0         0   
928609       34.0      26.0    EWR  MSY      1167          0         0   
851564       35.0      21.0    LGA  MSP      1020          0         0   

         AvarageSpeed  
893067         412.50  
869379         533.85  
303381         295.00  
1619984        443.89  
474335         394.07  
...               ...  
888638         422.61  
302880         543.83  
1306003        467.18  
928609         404.74  
851564         408.00  

[200000 rows x 19 columns]

#### Vuelo con retraso o dentro de hora

In [19]:
import numpy as np

vuelos['Delay_WithinTime']=np.where(vuelos['ArrDelay']>0, 'Delay', 'WithinTime')
vuelos

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
893067       6          10          2   1408.0        1255   1515.0   
869379       5          27          2   1026.0        1020   1614.0   
303381       2          11          1   1052.0        1045   1201.0   
1619984     10          23          4   1803.0        1750   2011.0   
474335       3          21          5   2209.0        2125   2251.0   
...        ...         ...        ...      ...         ...      ...   
888638       6           7          6   1202.0        1155   1413.0   
302880       2           5          2    932.0         810   1650.0   
1306003      8          15          5   1657.0        1648   2042.0   
928609       6          30          1   1331.0        1305   1546.0   
851564       5           9          5   1116.0        1055   1330.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  \
893067         1430            WN              127.0           155.0    112.0   
869379         1628            B6              228.0           248.0    195.0   
303381         1200            MQ               69.0            75.0     48.0   
1619984        2009            CO              248.0           259.0    221.0   
474335         2217            DL              102.0           112.0     81.0   
...             ...           ...                ...             ...      ...   
888638         1405            WN              131.0           130.0    115.0   
302880         1525            HA              318.0           315.0    282.0   
1306003        2018            YV              165.0           150.0    117.0   
928609         1512            XE              195.0           187.0    173.0   
851564         1255            AA              194.0           180.0    150.0   

         ArrDelay  DepDelay Origin Dest  Distance  Cancelled  Diverted  \
893067       45.0      73.0    STL  DEN       770          0         0   
869379      -14.0       6.0    LGB  ORD      1735          0         0   
303381        1.0       7.0    LAX  LAS       236          0         0   
1619984       2.0      13.0    IAH  SFO      1635          0         0   
474335       34.0      44.0    SLC  SMF       532          0         0   
...           ...       ...    ...  ...       ...        ...       ...   
888638        8.0       7.0    TPA  IAD       810          0         0   
302880       85.0      82.0    HNL  LAX      2556          0         0   
1306003      24.0       9.0    COS  ORD       911          0         0   
928609       34.0      26.0    EWR  MSY      1167          0         0   
851564       35.0      21.0    LGA  MSP      1020          0         0   

         AvarageSpeed Delay_WithinTime  
893067         412.50            Delay  
869379         533.85       WithinTime  
303381         295.00            Delay  
1619984        443.89            Delay  
474335         394.07            Delay  
...               ...              ...  
888638         422.61            Delay  
302880         543.83            Delay  
1306003        467.18            Delay  
928609         404.74            Delay  
851564         408.00            Delay  

[200000 rows x 20 columns]

#### (Creación de nueva columna) Tiempo de carreteo o movimiento del avión sobre tierra

In [20]:

vuelos['TaxiIn_and_TaxiOut']=(vuelos['ActualElapsedTime'] - vuelos['AirTime']).round(2)
vuelos


Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
893067       6          10          2   1408.0        1255   1515.0   
869379       5          27          2   1026.0        1020   1614.0   
303381       2          11          1   1052.0        1045   1201.0   
1619984     10          23          4   1803.0        1750   2011.0   
474335       3          21          5   2209.0        2125   2251.0   
...        ...         ...        ...      ...         ...      ...   
888638       6           7          6   1202.0        1155   1413.0   
302880       2           5          2    932.0         810   1650.0   
1306003      8          15          5   1657.0        1648   2042.0   
928609       6          30          1   1331.0        1305   1546.0   
851564       5           9          5   1116.0        1055   1330.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  ...  \
893067         1430            WN              127.0           155.0  ...   
869379         1628            B6              228.0           248.0  ...   
303381         1200            MQ               69.0            75.0  ...   
1619984        2009            CO              248.0           259.0  ...   
474335         2217            DL              102.0           112.0  ...   
...             ...           ...                ...             ...  ...   
888638         1405            WN              131.0           130.0  ...   
302880         1525            HA              318.0           315.0  ...   
1306003        2018            YV              165.0           150.0  ...   
928609         1512            XE              195.0           187.0  ...   
851564         1255            AA              194.0           180.0  ...   

         ArrDelay  DepDelay  Origin Dest Distance  Cancelled  Diverted  \
893067       45.0      73.0     STL  DEN      770          0         0   
869379      -14.0       6.0     LGB  ORD     1735          0         0   
303381        1.0       7.0     LAX  LAS      236          0         0   
1619984       2.0      13.0     IAH  SFO     1635          0         0   
474335       34.0      44.0     SLC  SMF      532          0         0   
...           ...       ...     ...  ...      ...        ...       ...   
888638        8.0       7.0     TPA  IAD      810          0         0   
302880       85.0      82.0     HNL  LAX     2556          0         0   
1306003      24.0       9.0     COS  ORD      911          0         0   
928609       34.0      26.0     EWR  MSY     1167          0         0   
851564       35.0      21.0     LGA  MSP     1020          0         0   

         AvarageSpeed  Delay_WithinTime TaxiIn_and_TaxiOut  
893067         412.50             Delay               15.0  
869379         533.85        WithinTime               33.0  
303381         295.00             Delay               21.0  
1619984        443.89             Delay               27.0  
474335         394.07             Delay               21.0  
...               ...               ...                ...  
888638         422.61             Delay               16.0  
302880         543.83             Delay               36.0  
1306003        467.18             Delay               48.0  
928609         404.74             Delay               22.0  
851564         408.00             Delay               44.0  

[200000 rows x 21 columns]

### d) Tabla de aerolíneas con más retrasos acumulados. Muestra de mayor a menor

In [21]:
#Creación de un dataframe solo con valores 'Delay' (retraso) de la columna ya creada anteriormente 'Delay_WithinTime'

solo_retrasos=vuelos.groupby('Delay_WithinTime').get_group('Delay')
solo_retrasos

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
893067       6          10          2   1408.0        1255   1515.0   
303381       2          11          1   1052.0        1045   1201.0   
1619984     10          23          4   1803.0        1750   2011.0   
474335       3          21          5   2209.0        2125   2251.0   
930772       6           1          7   2153.0        2130   2230.0   
...        ...         ...        ...      ...         ...      ...   
888638       6           7          6   1202.0        1155   1413.0   
302880       2           5          2    932.0         810   1650.0   
1306003      8          15          5   1657.0        1648   2042.0   
928609       6          30          1   1331.0        1305   1546.0   
851564       5           9          5   1116.0        1055   1330.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  ...  \
893067         1430            WN              127.0           155.0  ...   
303381         1200            MQ               69.0            75.0  ...   
1619984        2009            CO              248.0           259.0  ...   
474335         2217            DL              102.0           112.0  ...   
930772         2209            YV               37.0            39.0  ...   
...             ...           ...                ...             ...  ...   
888638         1405            WN              131.0           130.0  ...   
302880         1525            HA              318.0           315.0  ...   
1306003        2018            YV              165.0           150.0  ...   
928609         1512            XE              195.0           187.0  ...   
851564         1255            AA              194.0           180.0  ...   

         ArrDelay  DepDelay  Origin Dest Distance  Cancelled  Diverted  \
893067       45.0      73.0     STL  DEN      770          0         0   
303381        1.0       7.0     LAX  LAS      236          0         0   
1619984       2.0      13.0     IAH  SFO     1635          0         0   
474335       34.0      44.0     SLC  SMF      532          0         0   
930772       21.0      23.0     HNL  LIH      102          0         0   
...           ...       ...     ...  ...      ...        ...       ...   
888638        8.0       7.0     TPA  IAD      810          0         0   
302880       85.0      82.0     HNL  LAX     2556          0         0   
1306003      24.0       9.0     COS  ORD      911          0         0   
928609       34.0      26.0     EWR  MSY     1167          0         0   
851564       35.0      21.0     LGA  MSP     1020          0         0   

         AvarageSpeed  Delay_WithinTime TaxiIn_and_TaxiOut  
893067         412.50             Delay               15.0  
303381         295.00             Delay               21.0  
1619984        443.89             Delay               27.0  
474335         394.07             Delay               21.0  
930772         278.18             Delay               15.0  
...               ...               ...                ...  
888638         422.61             Delay               16.0  
302880         543.83             Delay               36.0  
1306003        467.18             Delay               48.0  
928609         404.74             Delay               22.0  
851564         408.00             Delay               44.0  

[177846 rows x 21 columns]

In [22]:
# Creación de un dataframe solo con valores 'Delay' (retraso) de la columna ya creada anteriormente 'Delay_WithinTime'
solo_retrasos=vuelos.groupby('Delay_WithinTime').get_group('Delay')


# Contabilización vuelos con retrasos por aerolínea
retrasos_aerol=pd.DataFrame(solo_retrasos.groupby('UniqueCarrier')['Delay_WithinTime'].value_counts())


# Contabilización vuelos totales por aerolínea
total_vuelos_aerol=pd.DataFrame(vuelos.groupby('UniqueCarrier')['Delay_WithinTime'].count())

# Porcentaje de vuelos retrasados con respecto a los vuelos totales por aerolínea
porcent_retr=(retrasos_aerol/total_vuelos_aerol)*100

porcent_retr.columns=['Porcentaje']
porcent_retr.sort_values(by='Porcentaje', ascending=False)



Porcentaje
UniqueCarrier Delay_WithinTime            
HA            Delay              95.910290
YV            Delay              94.453341
OH            Delay              93.289570
MQ            Delay              92.190305
EV            Delay              91.853433
OO            Delay              91.695075
F9            Delay              91.390952
FL            Delay              91.219178
NW            Delay              91.209461
XE            Delay              91.046165
AQ            Delay              90.588235
AA            Delay              89.889344
9E            Delay              89.685084
DL            Delay              88.097262
UA            Delay              87.878995
B6            Delay              87.156761
AS            Delay              86.564253
WN            Delay              85.763385
US            Delay              84.766560
CO            Delay              83.470106

### e)  Los vuelos más largos. Muestra de los 10 primeros para tener una buena visión de esta variable, pero sin llegar a ser excesiva. 
#### # Matizar que el ejercicio no pide una nueva tabla solo con los valores de la columna de 'Distancia' o alguna más, de lo contrario habría procedido como en el ejercicio anterior. Se entiende que presente los registros de los vuelos más largos.

In [23]:
vuelos.sort_values(by='Distance', ascending=False).head(10)

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
362852       2          20          3   1331.0        1320   1941.0   
1726009     11          21          5   1412.0        1335   2013.0   
175736       1          28          1   1359.0        1320   1928.0   
367026       2          11          1   1334.0        1320   1921.0   
1253766      7          25          5   1425.0        1335   1843.0   
1075813      6          14          6   1345.0        1335   1836.0   
1921778     12          31          3   1341.0        1335   1944.0   
556386       3          28          5   1329.0        1320   1848.0   
1618504     10          30          4   1408.0        1335   1838.0   
871090       5           7          3   1341.0        1335   1804.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  ...  \
362852         1920            CO              670.0           660.0  ...   
1726009        1935            CO              661.0           660.0  ...   
175736         1920            CO              629.0           660.0  ...   
367026         1920            CO              647.0           660.0  ...   
1253766        1810            CO              618.0           635.0  ...   
1075813        1810            CO              651.0           635.0  ...   
1921778        1935            CO              663.0           660.0  ...   
556386         1820            CO              679.0           660.0  ...   
1618504        1835            CO              630.0           660.0  ...   
871090         1810            CO              623.0           635.0  ...   

         ArrDelay  DepDelay  Origin Dest Distance  Cancelled  Diverted  \
362852       21.0      11.0     EWR  HNL     4962          0         0   
1726009      38.0      37.0     EWR  HNL     4962          0         0   
175736        8.0      39.0     EWR  HNL     4962          0         0   
367026        1.0      14.0     EWR  HNL     4962          0         0   
1253766      33.0      50.0     EWR  HNL     4962          0         0   
1075813      26.0      10.0     EWR  HNL     4962          0         0   
1921778       9.0       6.0     EWR  HNL     4962          0         0   
556386       28.0       9.0     EWR  HNL     4962          0         0   
1618504       3.0      33.0     EWR  HNL     4962          0         0   
871090       -6.0       6.0     EWR  HNL     4962          0         0   

         AvarageSpeed  Delay_WithinTime TaxiIn_and_TaxiOut  
362852         458.74             Delay               21.0  
1726009        478.65             Delay               39.0  
175736         494.55             Delay               27.0  
367026         478.65             Delay               25.0  
1253766        505.47             Delay               29.0  
1075813        482.53             Delay               34.0  
1921778        490.48             Delay               56.0  
556386         457.33             Delay               28.0  
1618504        493.73             Delay               27.0  
871090         502.06        WithinTime               30.0  

[10 rows x 21 columns]

In [24]:
pd.DataFrame(vuelos.groupby('UniqueCarrier')['Delay_WithinTime'].count())

Delay_WithinTime
UniqueCarrier                  
9E                         5303
AA                        19791
AQ                           85
AS                         4101
B6                         5754
CO                        10236
DL                        11762
EV                         8433
F9                         2962
FL                         7300
HA                          758
MQ                        14482
NW                         8202
OH                         5484
OO                        13787
UA                        14644
US                        10024
WN                        39279
XE                        10744
YV                         6869

### Los 10 vuelos con más retrasos. 
#### Procedimiento idem ejercicio anterior. 

In [25]:
vuelos.sort_values(by='ArrDelay', ascending=False).head(10)

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
163379       1           5          6    800.0        1045   1452.0   
836362       5           3          6    903.0        1131   1034.0   
836938       5          11          7    637.0         939    925.0   
682367       4           5          6    900.0        1200   1404.0   
1496940      9          21          7   1300.0        1633   1451.0   
1029214      6          18          3    711.0        1119   1153.0   
1599350     10          30          4    711.0        1205   1244.0   
1220767      7           2          3   1412.0        1925   1512.0   
697962       4           2          3    558.0        1120   1125.0   
142186       1          31          4    810.0        1345   1128.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  ...  \
163379         1327            AA              292.0            42.0  ...   
836362         1300            NW               91.0            89.0  ...   
836938         1222            NW              168.0           163.0  ...   
682367         1715            NW              184.0           195.0  ...   
1496940        1812            NW              111.0            99.0  ...   
1029214        1607            NW              162.0           168.0  ...   
1599350        1736            NW              213.0           211.0  ...   
1220767        2009            9E               60.0            44.0  ...   
697962         1659            AA              207.0           219.0  ...   
142186         1714            9E              138.0           149.0  ...   

         ArrDelay  DepDelay  Origin Dest Distance  Cancelled  Diverted  \
163379     1525.0    1275.0     EGE  MIA     1811          0         0   
836362     1294.0    1292.0     OKC  MEM      432          0         0   
836938     1263.0    1258.0     TPA  DTW      983          0         0   
682367     1249.0    1260.0     PDX  MSP     1426          0         0   
1496940    1239.0    1227.0     BWI  DTW      408          0         0   
1029214    1186.0    1192.0     DEN  DTW     1123          0         0   
1599350    1148.0    1146.0     SMF  MSP     1518          0         0   
1220767    1143.0    1127.0     FNT  DTW       56          0         0   
697962     1106.0    1118.0     LAS  ORD     1515          0         0   
142186     1094.0    1105.0     GTF  MSP      887          0         0   

         AvarageSpeed  Delay_WithinTime TaxiIn_and_TaxiOut  
163379         524.93             Delay               85.0  
836362         392.73             Delay               25.0  
836938         395.84             Delay               19.0  
682367         518.55             Delay               19.0  
1496940        388.57             Delay               48.0  
1029214        477.87             Delay               21.0  
1599350        508.83             Delay               34.0  
1220767        146.09             Delay               37.0  
697962         502.21             Delay               26.0  
142186         462.78             Delay               23.0  

[10 rows x 21 columns]

### Rutas más largas (seleccionadas las 10 primeras)

In [26]:
rutas_mas_larg = vuelos.drop_duplicates(subset = "Distance").sort_values(by = 'Distance', ascending = False).head(10)
rutas_mas_larg[['Origin', 'Dest', 'Distance']]

Origin Dest  Distance
362852     EWR  HNL      4962
808547     HNL  ATL      4502
1809925    ORD  HNL      4243
1323571    KOA  ORD      4213
956607     ORD  OGG      4184
1883745    MSP  HNL      3972
1414442    IAH  HNL      3904
1502411    DFW  HNL      3784
1385574    DFW  OGG      3711
1259550    ATL  ANC      3417

### f) Nuevos datos que considere relevantes

#### Cantidad de vuelos por mes

In [27]:
vuelos_mes = vuelos['Month'].value_counts().reset_index()
vuelos_mes.columns=['Número del mes', 'Cantidad de vuelos']
vuelos_mes.sort_values('Cantidad de vuelos', ascending=False)


Número del mes  Cantidad de vuelos
0                6               20871
1               12               20751
2                3               20595
3                2               19649
4                1               19113
5                7               19018
6                8               16871
7                4               16127
8                5               15800
9               11               10816
10              10               10713
11               9                9676

Los meses de Diciembre, Junio y Marzo son en los que más vuelos se realizan, mientras que en el mes de Septiembre es el que menos.

#### Cantidad vuelos por día del mes

In [28]:

vuelos_Dia_Mes=vuelos['DayofMonth'].value_counts().reset_index()
vuelos_Dia_Mes.columns=['Día del Mes', 'Cantidad de vuelos']
vuelos_Dia_Mes.sort_values('Cantidad de vuelos', ascending=False)


Día del Mes  Cantidad de vuelos
0            22                7419
1            21                7311
2             1                7043
3             7                6943
4            15                6864
5            20                6854
6            27                6802
7            23                6757
8            18                6674
9            24                6660
10           19                6642
11           14                6628
12           10                6618
13           11                6610
14            4                6582
15            2                6504
16            8                6463
17            6                6417
18           12                6369
19           17                6360
20           13                6350
21           26                6271
22           16                6249
23            9                6240
24           25                6239
25           28                6204
26            5                6141
27            3                6120
28           30                5847
29           29                5799
30           31                4020

Llama la atención como precisamente los últimos días de cada mes (31, 30 y 29) es cuando se realizan menos vuelos, en cambio el día 1 está entre los tres primeros.

#### Cantidad vuelos por día de la semana

In [29]:
vuelos_Dia_Sem=vuelos['DayOfWeek'].value_counts().reset_index()
vuelos_Dia_Sem.columns=['Día de la semana', 'Cantidad de vuelos']
vuelos_Dia_Sem.sort_values('Cantidad de vuelos', ascending=False)


Día de la semana  Cantidad de vuelos
0                 5               33552
1                 4               29950
2                 1               29817
3                 7               29498
4                 3               27236
5                 2               27087
6                 6               22860

En cuanto a los vuelos que se efectúan por día de la semana, el primer y último día laborables de la semana son los días que más vuelos se ejecutan. Por su parte, el sábado es el que menos.

## Ejercicio 3

#### Exportar el nuevo dataframe con las nuevas columnas añadidas a Excel

In [ ]:
vuelos_colum_relev.to_excel('Vuelos_análisis.xlsx')